### ISSUE TITLE
[Bug]: GPTQ w4a16 Quantization slower than FP16 (Please Help)

### ISSUE THREAD

**Issue by jvonrad** [vllm-project/vllm#21266]
</details>

## Problem

I quantized the llama3-8b myself using gptq, also downloaded the quantized hf version and also used vllm llm compressor and despite it saying that it uses the gptq_marlin kernels for all those I don't see a speedup in neither the awq nor the gptq quantized versions... (running on SM90 H100GPU with Cuda-12.2 and vllm-0.9.2 installed - i pulled newest main version) I also tried different input and output-len, as well as many more prompts like 1000, but its always pretty much the same like the float16 unquantized version.

## my setup:
(vllm) [gwb082@mlcbm002 vllm]$ python -c "import torch; print('Torch:', torch.__version__, '\nCUDA toolkit version:', torch.version.cuda, '\nCUDA available:', torch.cuda.is_available())"
Torch: 2.7.1+cu126 
CUDA toolkit version: 12.6 
CUDA available: True
(vllm) [gwb082@mlcbm002 vllm]$ vllm --version
INFO 07-20 01:56:52 [__init__.py:235] Automatically detected platform cuda.
0.9.2rc2.dev368+g2b504eb77.precompiled

## I am testing like this:

```
(vllm) [gwb082@mlcbm012 Meta-Llama-3-8B-Instruct-quantized.w4a16]$ vllm bench throughput     --model /home/geiger/gwb082/Jonathans_Thesis/compressed-models/quantized/Meta-Llama-3-8B-Instruct-quantized.w4a16     --input-len 32     --output-len 1     --enforce-eager     --load-format dummy
````
and despite logging:
"INFO 07-20 01:26:42 [gptq_marlin.py:266] Using MacheteLinearKernel for GPTQMarlinLinearMethod
INFO 07-20 01:26:42 [cuda.py:293] Using Flash Attention backend on V1 engine."

## Results
get this as results for quantized:
> Throughput: 1042.92 requests/s, 34382.06 total tokens/s, 1042.92 output tokens/s

and for the unquantized:
> Throughput: 1233.22 requests/s, 40625.98 total tokens/s, 1233.22 output tokens/s

## Entire Logs
This is the entire log for the quantized:

´´´´
(vllm) [gwb082@mlcbm012 Meta-Llama-3-8B-Instruct-quantized.w4a16]$ vllm bench throughput     --model /home/geiger/gwb082/Jonathans_Thesis/compressed-models/quantized/Meta-Llama-3-8B-Instruct-quantized.w4a16     --input-len 32     --output-len 1     --enforce-eager     --load-format dummy
INFO 07-20 23:35:44 [__init__.py:244] Automatically detected platform cuda.
When dataset path is not set, it will default to random dataset
INFO 07-20 23:35:49 [datasets.py:348] Sampling input_len from [31, 31] and output_len from [1, 1]
INFO 07-20 23:35:59 [config.py:823] This model supports multiple tasks: {'generate', 'reward', 'embed', 'score', 'classify'}. Defaulting to 'generate'.
INFO 07-20 23:35:59 [gptq_marlin.py:145] The model is convertible to gptq_marlin during runtime. Using gptq_marlin kernel.
INFO 07-20 23:35:59 [config.py:2195] Chunked prefill is enabled with max_num_batched_tokens=16384.
WARNING 07-20 23:35:59 [cuda.py:91] To see benefits of async output processing, enable CUDA graph. Since, enforce-eager is enabled, async output processor cannot be used
WARNING 07-20 23:36:03 [env_override.py:17] NCCL_CUMEM_ENABLE is set to 0, skipping override. This may increase memory overhead with cudagraph+allreduce: https://github.com/NVIDIA/nccl/issues/1234
INFO 07-20 23:36:06 [__init__.py:244] Automatically detected platform cuda.
INFO 07-20 23:36:09 [core.py:455] Waiting for init message from front-end.
INFO 07-20 23:36:09 [core.py:70] Initializing a V1 LLM engine (v0.9.1) with config: model='/home/geiger/gwb082/Jonathans_Thesis/compressed-models/quantized/Meta-Llama-3-8B-Instruct-quantized.w4a16', speculative_config=None, tokenizer='/home/geiger/gwb082/Jonathans_Thesis/compressed-models/quantized/Meta-Llama-3-8B-Instruct-quantized.w4a16', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config={}, tokenizer_revision=None, trust_remote_code=False, dtype=torch.float16, max_seq_len=8192, download_dir=None, load_format=LoadFormat.DUMMY, tensor_parallel_size=1, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=gptq_marlin, enforce_eager=True, kv_cache_dtype=auto,  device_config=cuda, decoding_config=DecodingConfig(backend='auto', disable_fallback=False, disable_any_whitespace=False, disable_additional_properties=False, reasoning_backend=''), observability_config=ObservabilityConfig(show_hidden_metrics_for_version=None, otlp_traces_endpoint=None, collect_detailed_traces=None), seed=0, served_model_name=/home/geiger/gwb082/Jonathans_Thesis/compressed-models/quantized/Meta-Llama-3-8B-Instruct-quantized.w4a16, num_scheduler_steps=1, multi_step_stream_outputs=True, enable_prefix_caching=True, chunked_prefill_enabled=True, use_async_output_proc=False, pooler_config=None, compilation_config={"level":0,"debug_dump_path":"","cache_dir":"","backend":"","custom_ops":[],"splitting_ops":[],"use_inductor":true,"compile_sizes":[],"inductor_compile_config":{"enable_auto_functionalized_v2":false},"inductor_passes":{},"use_cudagraph":true,"cudagraph_num_of_warmups":0,"cudagraph_capture_sizes":[],"cudagraph_copy_inputs":false,"full_cuda_graph":false,"max_capture_size":0,"local_cache_dir":null}
WARNING 07-20 23:36:09 [utils.py:2737] Methods determine_num_available_blocks,device_config,get_cache_block_size_bytes,initialize_cache not implemented in <vllm.v1.worker.gpu_worker.Worker object at 0x15533d03b650>
INFO 07-20 23:36:11 [parallel_state.py:1065] rank 0 in world size 1 is assigned as DP rank 0, PP rank 0, TP rank 0, EP rank 0
WARNING 07-20 23:36:11 [topk_topp_sampler.py:59] FlashInfer is not available. Falling back to the PyTorch-native implementation of top-p & top-k sampling. For the best performance, please install FlashInfer.
INFO 07-20 23:36:11 [gpu_model_runner.py:1595] Starting to load model /home/geiger/gwb082/Jonathans_Thesis/compressed-models/quantized/Meta-Llama-3-8B-Instruct-quantized.w4a16...
INFO 07-20 23:36:11 [gpu_model_runner.py:1600] Loading model from scratch...
INFO 07-20 23:36:11 [gptq_marlin.py:240] Using MacheteLinearKernel for GPTQMarlinLinearMethod
INFO 07-20 23:36:11 [cuda.py:252] Using Flash Attention backend on V1 engine.
INFO 07-20 23:36:12 [gpu_model_runner.py:1624] Model loading took 5.3507 GiB and 0.549619 seconds
INFO 07-20 23:36:14 [gpu_worker.py:227] Available KV cache memory: 61.00 GiB
INFO 07-20 23:36:14 [kv_cache_utils.py:715] GPU KV cache size: 499,696 tokens
INFO 07-20 23:36:14 [kv_cache_utils.py:719] Maximum concurrency for 8,192 tokens per request: 61.00x
INFO 07-20 23:36:14 [core.py:171] init engine (profile, create kv cache, warmup model) took 1.91 seconds
Adding requests: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 4628.27it/s]
Processed prompts: 100%|███████████████████████████████████████████| 1000/1000 [00:00<00:00, 1350.04it/s, est. speed input: 43159.07 toks/s, output: 1350.11 toks/s]
[rank0]:[W720 23:36:16.398396939 ProcessGroupNCCL.cpp:1476] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
Throughput: 1042.92 requests/s, 34382.06 total tokens/s, 1042.92 output tokens/s
Total num prompt tokens:  31967
Total num output tokens:  1000
´´´´

and this is the entire log for the not-quantized:

´´´´
(vllm) [gwb082@mlcbm012 llama-3-8b-wanda-2.4_clean]$ vllm bench throughput \
>     --model /home/geiger/gwb082/LLMs/llama-3/Meta-Llama-3-8B \
>     --input-len 32 \
>     --output-len 1 \
>     --enforce-eager \
>     --load-format dummy
INFO 07-20 23:32:26 [__init__.py:244] Automatically detected platform cuda.
When dataset path is not set, it will default to random dataset
INFO 07-20 23:32:32 [datasets.py:348] Sampling input_len from [31, 31] and output_len from [1, 1]
INFO 07-20 23:32:42 [config.py:823] This model supports multiple tasks: {'score', 'reward', 'generate', 'classify', 'embed'}. Defaulting to 'generate'.
INFO 07-20 23:32:42 [config.py:2195] Chunked prefill is enabled with max_num_batched_tokens=16384.
WARNING 07-20 23:32:42 [cuda.py:91] To see benefits of async output processing, enable CUDA graph. Since, enforce-eager is enabled, async output processor cannot be used
WARNING 07-20 23:32:46 [env_override.py:17] NCCL_CUMEM_ENABLE is set to 0, skipping override. This may increase memory overhead with cudagraph+allreduce: https://github.com/NVIDIA/nccl/issues/1234
INFO 07-20 23:32:49 [__init__.py:244] Automatically detected platform cuda.
INFO 07-20 23:32:52 [core.py:455] Waiting for init message from front-end.
INFO 07-20 23:32:52 [core.py:70] Initializing a V1 LLM engine (v0.9.1) with config: model='/home/geiger/gwb082/LLMs/llama-3/Meta-Llama-3-8B', speculative_config=None, tokenizer='/home/geiger/gwb082/LLMs/llama-3/Meta-Llama-3-8B', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config={}, tokenizer_revision=None, trust_remote_code=False, dtype=torch.bfloat16, max_seq_len=8192, download_dir=None, load_format=LoadFormat.DUMMY, tensor_parallel_size=1, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=None, enforce_eager=True, kv_cache_dtype=auto,  device_config=cuda, decoding_config=DecodingConfig(backend='auto', disable_fallback=False, disable_any_whitespace=False, disable_additional_properties=False, reasoning_backend=''), observability_config=ObservabilityConfig(show_hidden_metrics_for_version=None, otlp_traces_endpoint=None, collect_detailed_traces=None), seed=0, served_model_name=/home/geiger/gwb082/LLMs/llama-3/Meta-Llama-3-8B, num_scheduler_steps=1, multi_step_stream_outputs=True, enable_prefix_caching=True, chunked_prefill_enabled=True, use_async_output_proc=False, pooler_config=None, compilation_config={"level":0,"debug_dump_path":"","cache_dir":"","backend":"","custom_ops":[],"splitting_ops":[],"use_inductor":true,"compile_sizes":[],"inductor_compile_config":{"enable_auto_functionalized_v2":false},"inductor_passes":{},"use_cudagraph":true,"cudagraph_num_of_warmups":0,"cudagraph_capture_sizes":[],"cudagraph_copy_inputs":false,"full_cuda_graph":false,"max_capture_size":0,"local_cache_dir":null}
WARNING 07-20 23:32:52 [utils.py:2737] Methods determine_num_available_blocks,device_config,get_cache_block_size_bytes,initialize_cache not implemented in <vllm.v1.worker.gpu_worker.Worker object at 0x15533d1efe90>
INFO 07-20 23:32:54 [parallel_state.py:1065] rank 0 in world size 1 is assigned as DP rank 0, PP rank 0, TP rank 0, EP rank 0
WARNING 07-20 23:32:54 [topk_topp_sampler.py:59] FlashInfer is not available. Falling back to the PyTorch-native implementation of top-p & top-k sampling. For the best performance, please install FlashInfer.
INFO 07-20 23:32:54 [gpu_model_runner.py:1595] Starting to load model /home/geiger/gwb082/LLMs/llama-3/Meta-Llama-3-8B...
INFO 07-20 23:32:54 [gpu_model_runner.py:1600] Loading model from scratch...
INFO 07-20 23:32:54 [cuda.py:252] Using Flash Attention backend on V1 engine.
INFO 07-20 23:32:55 [gpu_model_runner.py:1624] Model loading took 14.9596 GiB and 0.153384 seconds
INFO 07-20 23:32:56 [gpu_worker.py:227] Available KV cache memory: 51.39 GiB
INFO 07-20 23:32:56 [kv_cache_utils.py:715] GPU KV cache size: 420,976 tokens
INFO 07-20 23:32:56 [kv_cache_utils.py:719] Maximum concurrency for 8,192 tokens per request: 51.39x
INFO 07-20 23:32:56 [core.py:171] init engine (profile, create kv cache, warmup model) took 1.76 seconds
Adding requests: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 4030.74it/s]
Processed prompts: 100%|███████████████████████████████████████████| 1000/1000 [00:00<00:00, 1785.82it/s, est. speed input: 57052.35 toks/s, output: 1786.05 toks/s]
[rank0]:[W720 23:32:58.482502636 ProcessGroupNCCL.cpp:1476] Warning: WARNING: destroy_process_group() was not called before program exit, which can leak resources. For more info, please see https://pytorch.org/docs/stable/distributed.html#shutdown (function operator())
Throughput: 1233.22 requests/s, 40625.98 total tokens/s, 1233.22 output tokens/s
Total num prompt tokens:  31943
Total num output tokens:  1000
´´´´

Please help.

### Your current environment (if you think it is necessary)

```text
The output of `python collect_env.py`
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by jeejeelee** [comment]
Have you tried removing --enforce-eager?

**Comment by jvonrad** [comment]
> Have you tried removing --enforce-eager?

Yes, unfortunately still the quantized version is a bit slower. I feel like A100 and H100 are just too optimized for Bf16 matrix mult, so the quantization doesn't help actually?

**Comment by JINO-ROHIT** [comment]
were you able to solve this?
